#### 1. Import pandas library

In [1]:
import pandas as pd

#### 2. Import pymysql and sqlalchemy as you have learnt in the lesson of importing/exporting data 


In [3]:
!pip3 install SQLAlchemy
!pip3 install PyMySQL
!pip3 install mysql-connector

  Using cached mysql-connector-2.2.9.tar.gz (11.9 MB)
  Created wheel for mysql-connector: filename=mysql_connector-2.2.9-cp38-cp38-win_amd64.whl size=247946 sha256=2d5d4627734c46cf36579fcae1c38bd80c27a0572cf460f54aa2043164487d81
  Stored in directory: c:\users\irock\appdata\local\pip\cache\wheels\57\e4\98\5feafb5c393dd2540e44b064a6f95832990d543e5b4f53ea8f
Successfully built mysql-connector


In [4]:
import pymysql
from sqlalchemy import create_engine

#### 3. Create a mysql engine to set the connection to the server. Check the connection details in [this link](https://relational.fit.cvut.cz/search?tableCount%5B%5D=0-10&tableCount%5B%5D=10-30&dataType%5B%5D=Numeric&databaseSize%5B%5D=KB&databaseSize%5B%5D=MB)

In [5]:
#https://relational.fit.cvut.cz/dataset/Stats
engine = create_engine('mysql+pymysql://guest:relational@relational.fit.cvut.cz/stats')

In [28]:
engine.table_names()

['badges',
 'comments',
 'postHistory',
 'postLinks',
 'posts',
 'tags',
 'users',
 'votes']

#### 4. Import the users table 

In [48]:
#https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql.html
#pandas.read_sql(sql, con, index_col='None', coerce_float='True', params='None', parse_dates='None', columns='None', chunksize: None = 'None')
users_table = pd.read_sql('SELECT * FROM users', engine)

In [49]:
users_table.head()

Id  Reputation        CreationDate   DisplayName      LastAccessDate  \
0  -1           1 2010-07-19 06:55:26     Community 2010-07-19 06:55:26   
1   2         101 2010-07-19 14:01:36  Geoff Dalgas 2013-11-12 22:07:23   
2   3         101 2010-07-19 15:34:50  Jarrod Dixon 2014-08-08 06:42:58   
3   4         101 2010-07-19 19:03:27        Emmett 2014-01-02 09:31:02   
4   5        6792 2010-07-19 19:03:57         Shane 2014-08-13 00:23:47   

                       WebsiteUrl            Location  \
0  http://meta.stackexchange.com/  on the server farm   
1        http://stackoverflow.com       Corvallis, OR   
2        http://stackoverflow.com        New York, NY   
3    http://minesweeperonline.com   San Francisco, CA   
4         http://www.statalgo.com        New York, NY   

                                             AboutMe  Views  UpVotes  \
0  <p>Hi, I'm not really a person.</p>\n\n<p>I'm ...      0     5007   
1  <p>Developer on the StackOverflow team.  Find ...     25        3   
2  <p><a href="http://blog.stackoverflow.com/2009...     22       19   
3  <p>currently at a startup in SF</p>\n\n<p>form...     11        0   
4  <p>Quantitative researcher focusing on statist...   1145      662   

   DownVotes  AccountId   Age                     ProfileImageUrl  
0       1920         -1   NaN                                None  
1          0          2  37.0                                None  
2          0          3  35.0                                None  
3          0       1998  28.0  http://i.stack.imgur.com/d1oHX.jpg  
4          5      54503  35.0                                None

#### 5. Rename Id column to userId

In [50]:
users_table = users_table.rename(columns={'Id':'userid'})

In [51]:
users_table.head()

userid  Reputation        CreationDate   DisplayName      LastAccessDate  \
0      -1           1 2010-07-19 06:55:26     Community 2010-07-19 06:55:26   
1       2         101 2010-07-19 14:01:36  Geoff Dalgas 2013-11-12 22:07:23   
2       3         101 2010-07-19 15:34:50  Jarrod Dixon 2014-08-08 06:42:58   
3       4         101 2010-07-19 19:03:27        Emmett 2014-01-02 09:31:02   
4       5        6792 2010-07-19 19:03:57         Shane 2014-08-13 00:23:47   

                       WebsiteUrl            Location  \
0  http://meta.stackexchange.com/  on the server farm   
1        http://stackoverflow.com       Corvallis, OR   
2        http://stackoverflow.com        New York, NY   
3    http://minesweeperonline.com   San Francisco, CA   
4         http://www.statalgo.com        New York, NY   

                                             AboutMe  Views  UpVotes  \
0  <p>Hi, I'm not really a person.</p>\n\n<p>I'm ...      0     5007   
1  <p>Developer on the StackOverflow team.  Find ...     25        3   
2  <p><a href="http://blog.stackoverflow.com/2009...     22       19   
3  <p>currently at a startup in SF</p>\n\n<p>form...     11        0   
4  <p>Quantitative researcher focusing on statist...   1145      662   

   DownVotes  AccountId   Age                     ProfileImageUrl  
0       1920         -1   NaN                                None  
1          0          2  37.0                                None  
2          0          3  35.0                                None  
3          0       1998  28.0  http://i.stack.imgur.com/d1oHX.jpg  
4          5      54503  35.0                                None

#### 6. Import the posts table. 

In [52]:
posts_table = pd.read_sql_query('SELECT * FROM posts', engine)

In [53]:
posts_table.head()

Id  PostTypeId  AcceptedAnswerId         CreaionDate  Score  ViewCount  \
0   1           1              15.0 2010-07-19 19:12:12     23     1278.0   
1   2           1              59.0 2010-07-19 19:12:57     22     8198.0   
2   3           1               5.0 2010-07-19 19:13:28     54     3613.0   
3   4           1             135.0 2010-07-19 19:13:31     13     5224.0   
4   5           2               NaN 2010-07-19 19:14:43     81        NaN   

                                                Body  OwnerUserId  \
0  <p>How should I elicit prior distributions fro...          8.0   
1  <p>In many different statistical methods there...         24.0   
2  <p>What are some valuable Statistical Analysis...         18.0   
3  <p>I have two groups of data.  Each with a dif...         23.0   
4  <p>The R-project</p>\n\n<p><a href="http://www...         23.0   

      LasActivityDate                                              Title  ...  \
0 2010-09-15 21:08:26                      Eliciting priors from experts  ...   
1 2012-11-12 09:21:54                                 What is normality?  ...   
2 2013-05-27 14:48:36  What are some valuable Statistical Analysis op...  ...   
3 2010-09-08 03:00:19  Assessing the significance of differences in d...  ...   
4 2010-07-19 19:21:15                                               None  ...   

  AnswerCount  CommentCount  FavoriteCount  LastEditorUserId  \
0         5.0             1           14.0               NaN   
1         7.0             1            8.0              88.0   
2        19.0             4           36.0             183.0   
3         5.0             2            2.0               NaN   
4         NaN             3            NaN              23.0   

         LastEditDate  CommunityOwnedDate ParentId  ClosedDate  \
0                 NaT                 NaT      NaN         NaT   
1 2010-08-07 17:56:44                 NaT      NaN         NaT   
2 2011-02-12 05:50:03 2010-07-19 19:13:28      NaN         NaT   
3                 NaT                 NaT      NaN         NaT   
4 2010-07-19 19:21:15 2010-07-19 19:14:43      3.0         NaT   

  OwnerDisplayName LastEditorDisplayName  
0             None                  None  
1             None                  None  
2             None                  None  
3             None                  None  
4             None                  None  

[5 rows x 21 columns]

#### 7. Rename Id column to postId and OwnerUserId to userId

In [54]:
posts_table = posts_table.rename(columns={'Id':'postId',
                                          'OwnerUserId':'userid'})

In [55]:
posts_table.head()

postId  PostTypeId  AcceptedAnswerId         CreaionDate  Score  ViewCount  \
0       1           1              15.0 2010-07-19 19:12:12     23     1278.0   
1       2           1              59.0 2010-07-19 19:12:57     22     8198.0   
2       3           1               5.0 2010-07-19 19:13:28     54     3613.0   
3       4           1             135.0 2010-07-19 19:13:31     13     5224.0   
4       5           2               NaN 2010-07-19 19:14:43     81        NaN   

                                                Body  userid  \
0  <p>How should I elicit prior distributions fro...     8.0   
1  <p>In many different statistical methods there...    24.0   
2  <p>What are some valuable Statistical Analysis...    18.0   
3  <p>I have two groups of data.  Each with a dif...    23.0   
4  <p>The R-project</p>\n\n<p><a href="http://www...    23.0   

      LasActivityDate                                              Title  ...  \
0 2010-09-15 21:08:26                      Eliciting priors from experts  ...   
1 2012-11-12 09:21:54                                 What is normality?  ...   
2 2013-05-27 14:48:36  What are some valuable Statistical Analysis op...  ...   
3 2010-09-08 03:00:19  Assessing the significance of differences in d...  ...   
4 2010-07-19 19:21:15                                               None  ...   

  AnswerCount  CommentCount  FavoriteCount  LastEditorUserId  \
0         5.0             1           14.0               NaN   
1         7.0             1            8.0              88.0   
2        19.0             4           36.0             183.0   
3         5.0             2            2.0               NaN   
4         NaN             3            NaN              23.0   

         LastEditDate  CommunityOwnedDate ParentId  ClosedDate  \
0                 NaT                 NaT      NaN         NaT   
1 2010-08-07 17:56:44                 NaT      NaN         NaT   
2 2011-02-12 05:50:03 2010-07-19 19:13:28      NaN         NaT   
3                 NaT                 NaT      NaN         NaT   
4 2010-07-19 19:21:15 2010-07-19 19:14:43      3.0         NaT   

  OwnerDisplayName LastEditorDisplayName  
0             None                  None  
1             None                  None  
2             None                  None  
3             None                  None  
4             None                  None  

[5 rows x 21 columns]

#### 8. Define new dataframes for users and posts with the following selected columns:
    **users columns**: userId, Reputation,Views,UpVotes,DownVotes
    **posts columns**: postId, Score,userID,ViewCount,CommentCount

In [56]:
users_new = users_table[['userid','Reputation','Views','UpVotes','DownVotes']]
users_new.head()

userid  Reputation  Views  UpVotes  DownVotes
0      -1           1      0     5007       1920
1       2         101     25        3          0
2       3         101     22       19          0
3       4         101     11        0          0
4       5        6792   1145      662          5

In [59]:
posts_new = posts_table[['postId','Score','userid','ViewCount','CommentCount']]
posts_new.head()

postId  Score  userid  ViewCount  CommentCount
0       1     23     8.0     1278.0             1
1       2     22    24.0     8198.0             1
2       3     54    18.0     3613.0             4
3       4     13    23.0     5224.0             2
4       5     81    23.0        NaN             3

#### 8. Merge both dataframes, users and posts. 
You will need to make a [merge](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) of posts and users dataframes.

In [96]:
merged = users_new.merge(posts_new, how='inner', on='userid')

In [97]:
merged

userid  Reputation  Views  UpVotes  DownVotes  postId  Score  \
0          -1           1      0     5007       1920    2175      0   
1          -1           1      0     5007       1920    8576      0   
2          -1           1      0     5007       1920    8578      0   
3          -1           1      0     5007       1920    8981      0   
4          -1           1      0     5007       1920    8982      0   
...       ...         ...    ...      ...        ...     ...    ...   
90579   55734           1      0        0          0  115352      0   
90580   55738          11      0        0          0  115360      2   
90581   55742           6      0        0          0  115366      1   
90582   55744           6      1        0          0  115370      1   
90583   55746         106      1        0          0  115376      1   

       ViewCount  CommentCount  
0            NaN             0  
1            NaN             0  
2            NaN             0  
3            NaN             0  
4            NaN             0  
...          ...           ...  
90579       16.0             0  
90580       40.0             4  
90581       17.0             0  
90582       13.0             2  
90583        5.0             2  

[90584 rows x 9 columns]

#### 9. How many missing values do you have in your merged dataframe? On which columns?

In [98]:
#48396 missing values on ViewCount column
merged.isnull().sum()

userid              0
Reputation          0
Views               0
UpVotes             0
DownVotes           0
postId              0
Score               0
ViewCount       48396
CommentCount        0
dtype: int64

#### 10. You will need to make something with missing values.  Will you clean or filling them? Explain. 
**Remember** to check the results of your code before passing to the next step

In [99]:
merged.shape

(90584, 9)

In [100]:
print('Porcentaje de datos faltantes %.2f%%' %((merged['ViewCount'].isnull().sum()/merged.shape[0])*100))

Porcentaje de datos faltantes 53.43%


Just over half of the data in ViewCount is missing (53.43%) so before taking the decision to drop the column I think it would be important to discuss about why this information is missing and evaluate how important is this column to analyze the data.
By now I will change the NaN values to 0

In [101]:
merged['ViewCount'].fillna(0, inplace = True)

In [102]:
merged

userid  Reputation  Views  UpVotes  DownVotes  postId  Score  \
0          -1           1      0     5007       1920    2175      0   
1          -1           1      0     5007       1920    8576      0   
2          -1           1      0     5007       1920    8578      0   
3          -1           1      0     5007       1920    8981      0   
4          -1           1      0     5007       1920    8982      0   
...       ...         ...    ...      ...        ...     ...    ...   
90579   55734           1      0        0          0  115352      0   
90580   55738          11      0        0          0  115360      2   
90581   55742           6      0        0          0  115366      1   
90582   55744           6      1        0          0  115370      1   
90583   55746         106      1        0          0  115376      1   

       ViewCount  CommentCount  
0            0.0             0  
1            0.0             0  
2            0.0             0  
3            0.0             0  
4            0.0             0  
...          ...           ...  
90579       16.0             0  
90580       40.0             4  
90581       17.0             0  
90582       13.0             2  
90583        5.0             2  

[90584 rows x 9 columns]

In [103]:
merged.isnull().sum()

userid          0
Reputation      0
Views           0
UpVotes         0
DownVotes       0
postId          0
Score           0
ViewCount       0
CommentCount    0
dtype: int64

#### 11. Adjust the data types in order to avoid future issues. Which ones should be changed? 

In [104]:
merged.dtypes

userid            int64
Reputation        int64
Views             int64
UpVotes           int64
DownVotes         int64
postId            int64
Score             int64
ViewCount       float64
CommentCount      int64
dtype: object

In [105]:
merged['ViewCount'].sum() #as the sum does not have decimals I will change the column Viewcount to 'int64' to prevent issues

23484210.0

In [110]:
import numpy as np
#merged['ViewCount'] = merged['ViewCount'].astype(int)
merged['ViewCount']=merged['ViewCount'].astype(np.int64)
merged.dtypes

userid          int64
Reputation      int64
Views           int64
UpVotes         int64
DownVotes       int64
postId          int64
Score           int64
ViewCount       int64
CommentCount    int64
dtype: object

#### Bonus: Identify extreme values in your merged dataframe as you have learned in class, create a dataframe called outliers with the same columns as our data set and calculate the bounds. The values of the outliers dataframe will be the values of the merged_df that fall outside that bounds. You will need to save your outliers dataframe to a csv file on your-code folder.

In [112]:
stats = merged.describe().T
stats['IQR'] = stats['75%'] - stats['50%']
stats

count          mean           std   min       25%      50%  \
userid        90584.0  16546.764727  15273.367108  -1.0   3437.00  11032.0   
Reputation    90584.0   6282.395412  15102.268670   1.0     60.00    396.0   
Views         90584.0   1034.245176   2880.074012   0.0      5.00     45.0   
UpVotes       90584.0    734.315718   2050.869327   0.0      1.00     22.0   
DownVotes     90584.0     33.273249    134.936435   0.0      0.00      0.0   
postId        90584.0  56539.080522  33840.307529   1.0  26051.75  57225.5   
Score         90584.0      2.780767      4.948922 -19.0      1.00      2.0   
ViewCount     90584.0    259.253400   1632.261405   0.0      0.00      0.0   
CommentCount  90584.0      1.894650      2.638704   0.0      0.00      1.0   

                   75%       max       IQR  
userid        27700.00   55746.0  16668.00  
Reputation     4460.00   87393.0   4064.00  
Views           514.25   20932.0    469.25  
UpVotes         283.00   11442.0    261.00  
DownVotes         8.00    1920.0      8.00  
postId        86145.25  115378.0  28919.75  
Score             3.00     192.0      1.00  
ViewCount       111.00  175495.0    111.00  
CommentCount      3.00      45.0      2.00

In [116]:
outliers = pd.DataFrame(columns=merged.columns)

for col in stats.index:
    iqr = stats.at[col, 'IQR']
    cutoff = iqr * 1.5
    lower = stats.at[col,'25%'] - cutoff
    upper = stats.at[col,'75%'] + cutoff
    results = merged[(merged[col] < lower) | (merged[col] > upper)].copy()
    results['Outlier'] = col
    outliers = outliers.append(results)
    
outliers

userid Reputation Views UpVotes DownVotes  postId Score ViewCount  \
88743  52703         21     0       0         0  109206     4        87   
88744  52705         11     1       0         0  111037     0         0   
88745  52705         11     1       0         0  111039     1         0   
88746  52706          8     0       0         0  109110     1        35   
88747  52708         13     2       0         0  109189     1         0   
...      ...        ...   ...     ...       ...     ...   ...       ...   
90482  55541          1     7       0         0  115109     0        22   
90491  55557          6     3       0         0  115020     1        77   
90535  55635        109     0       0         0  115172     1        40   
90539  55648          6     2       0         0  115182     1        18   
90561  55689          6     0       0         0  115270     1        28   

      CommentCount       Outlier  
88743            7        userid  
88744            0        userid  
88745            2        userid  
88746            8        userid  
88747            3        userid  
...            ...           ...  
90482            7  CommentCount  
90491            9  CommentCount  
90535            7  CommentCount  
90539            7  CommentCount  
90561            7  CommentCount  

[97795 rows x 10 columns]

In [117]:
outliers.to_csv('outliers.csv', index = False, sep=',')